GVFS/GVFS.Virtualization/Projection/Readme.md

```diff
@@ -0,0 +1,181 @@
+
+Class only used for file systems that support file mode since that is in the git index and is needed when the file is created on disk.
+
+### `PoolAllocationMultipliers`
+
+Class used to hold the multipliers that are applied to the various pools in the code.  These numbers come from running with various sized repos and determining what was best for keeping the pools at reasonable sizes.
+
+### `ObjectPool`
+
+Class that is a generic pool of some type of object that will dynamically grow or can be shrunk to free objects when too many get allocated.  All objects for the pool are created at the time the pool is expanded.  The `LazyUTF8String.BytePool` is a specialized pool to allow the use of a pointer into the allocated `byte[]`.
+
+### `FolderEntryData`
+
+Abstract base class for data about an item that is in a folder.  Contains the name and a flag for whether the entry is a folder.  `FolderData` and `FileData` are the derived classes for this class.
+
+### `FolderData`
+
+Class containing the data about a folder in the projection.  Includes the child entries as a `SortedFolderEntries` object, a flag to indicate the children's sizes have been populated, and a flag to indicate if the folder should be included in the projection (This is when using sparse mode).
+
+### `FileData`
+
+Class containing the data about a file in the projection.  Includes the size and the SHA1.  The SHA1 is stored as 2 `ulong` and an `uint` for performance and memory usage.
+
+### `LazyUTF8String`
+
+Class used to keep track of the string from the index that is in the `BytePool` and converts from the `BytePool` to a `string` on when needed by either calling the `GetString` method or `Compare` when one string is not all ASCII.
+
+### `SortedFolderEntries`
+
+Class used to keep the list entries for a folder, either `FolderData` or `FileData` objects) in sorted order.  This class keeps the static pool of both `FolderData` and `FileData` objects for reuse.
+
+Couple of things to note:
+
+1. This is using the `Compare` method of the `LazyUTF8String` class for a performance optimization since most of the time the paths in the index are ASCII and the code can do byte by byte comparison and __not__ have to convert to a `string` object and then compare which is a performance and memory hit.
+
+2. When getting the index of the name in the sorted entries it will return the bitwise complement of the index where the item should be inserted.  This was done to avoid making one call to determine if the name exists and a second call to get the index for insertion.
+
+### `SparseFolderData`
+
+Class used to keep the sparse folder information.  It contains a flag for whether the folder should be recursed into for projection, the depth of the folder, and the children in a name, data `Dictionary<string, SparseFolderData>`.
+
+```
+Example:
+
+For sparse folder entries of:
+foo/example
+other
+
+The SparseFolderData would be:
+root
+Children:
+|- foo (IsRecursive = false, Depth = 0)

```

Comment:
`IsRecursive` is false here because there can be no children of `foo/` in this particular example, just `foo/example`? Might be worth explaining just a touch more about the sparse model. For example, does root get special treatment?